### ISSUE TITLE
Any plans to support ML-DSA and ML-KEM?

### ISSUE THREAD

**Issue by Hussain1811** [Mbed-TLS/mbedtls#9803]
### Suggested enhancement

Add ML-DSA and ML-KEM post-quantum algorithm support

### Justification

Mbed TLS needs this because other security solutions which depend on MbeTLS can benefit from PQC support

**Comment by Hussain1811** [comment]
Thank you @tom-daubney-arm 
I've sent a mail to the mailing list

**Comment by yanesca** [comment]
Adding post quantum crypto is on our [roadmap](https://mbed-tls.readthedocs.io/en/latest/project/roadmap/) but we don't have any timelines yet.

**Comment by Hussain1811** [comment]
So, this GitHub issue can be kept open as "enhancement" tag has been added?

**Comment by tom-daubney-arm** [comment]
@Hussain1811 Yep that is correct as I mentioned in the first reply I will keep this open as it logs your interest in the algorithms. However I just thought that you would be more likely to get a better answer to your question if it went to the mailing list. @yanesca has answered this here though for you by posting the roadmap link.

Apologies if that was confusing.

**Comment by aparcar** [comment]
The [last emails regarding this topic](https://lists.trustedfirmware.org/archives/list/mbed-tls@lists.trustedfirmware.org/message/MICCDHUVHLB3PG3VY6VO2NYP7EJZ5OIF/) suggest 2025Q3 as a possible point to move forward. I don't find any newer developments. Does anyone following knows of more recent developments?

**Comment by gilles-peskine-arm** [comment]
We're working on it, but with very little bandwidth, unfortunately. We're currently integrating [mldsa-native](https://github.com/pq-code-package/mldsa-native) and then we'll integrate [mlkem-native](https://github.com/pq-code-package/mlkem-native). Right now, we have MLDSA at the driver dispatch layer (so suitable secure boot, but not general applications).

I hope we can get MLDSA at the PSA API layer in Q3, but that's only a hope, not a commitment. It'll depend on what other priorities come up and on our security workload (currently very high).

**Comment by mczraf** [comment]
Hi folks, I'm Rafael Misoczki from Meta. We were wondering if there's any progress in enabling PQC in mbedtls upstream. We have some prototypes currently using mbedtls for which we would like to protect considering the quantum threat model, in particular what concerns store-now-decrypt-later attacks (thus MLKEM support would be higher priority). Anything you could share at his point? Thank you for the great work.

**Comment by gilles-peskine-arm** [comment]
> I hope we can get MLDSA at the PSA API layer in Q3, but that's only a hope, not a commitment. It'll depend on what other priorities come up and on our security workload (currently very high).

Now 2026Q3 has started, our security workload has thankfully gone slightly down. But we're a European team and holidays mean that we don't have more bandwidth for feature development. So we won't be working on the MLDSA API layer this quarter.

As usual, our biggest bottleneck is review. Even more so nowadays, now that LLM can produce halfway decent code with sufficient guidance, but we don't yet trust them enough for review. So the best way to help is if more people can join the team on a long-term basis (but part-time is fine), so that we can rely on them as trusted reviewers.

**Comment by reneleonhardt** [comment]
Would it be possible to support contributors at least by enabling AI reviews, before maintainers review the potentially improved PR?
https://github.com/marketplace/coderabbitai

**Comment by davidhorstmann-arm** [comment]
> Would it be possible to support contributors at least by enabling AI reviews, before maintainers review the potentially improved PR?

We occasionally use AI to assist us in our normal reviews but we don't usually do initial AI reviews.

The reason is that we want to make sure we have the bandwidth to receive a contribution before we start reviewing it. If we just do an AI review and the contributor does a lot of rework but then we never actually do a human review, that wastes the contributor's time and is generally unhelpful.

**Comment by reneleonhardt** [comment]
Understood.

Could you enable manual AI review requests, for example CodeRabbit?
Some contributors benefit from that like me.

**Comment by gilles-peskine-arm** [comment]
Feel free to use the code reviews of your choice. However, beware that they may give bad advice. We have tried GitHub copilot and OpenAI Codex, and gotten mixed results. At this point, we are not willing to endorse any particular AI review tool.

We are working on `AGENTS.md` which should help, but we don't have anything yet that's ready for public consumption. Both human and AI reviewers should be aware of the [review guidelines](https://mbed-tls.readthedocs.io/en/latest/reviews/) and the [coding standards](https://mbed-tls.readthedocs.io/en/latest/kb/development/mbedtls-coding-standards/).
